libraries/AP_HAL_Linux/UARTDriver.cpp

```diff
@@ -553,22 +450,16 @@ void UARTDriver::_timer_tick(void)
     }
 
     // try to fill the read buffer
-    uint16_t _head;
-    n = BUF_SPACE(_readbuf);
-    if (n > 0) {
-        uint16_t n1 = _readbuf_size - _readbuf_tail;
-        if (n1 >= n) {
-            // one read will do
-            assert(_readbuf_tail+n <= _readbuf_size);
-            _read_fd(&_readbuf[_readbuf_tail], n);
-        } else {
-            assert(_readbuf_tail+n1 <= _readbuf_size);
-            int ret = _read_fd(&_readbuf[_readbuf_tail], n1);
-            if (ret == n1 && n > n1) {
-                assert(_readbuf_tail+(n-n1) <= _readbuf_size);
-                _read_fd(&_readbuf[_readbuf_tail], n - n1);
-            }
-        }
+    int ret;
+    ByteBuffer::IoVec vec[2];
+
+    const auto n_vec = _readbuf.reserve(vec, _readbuf.space());
+    for (int i = 0; i < n_vec; i++) {
+        ret = _read_fd(vec[i].data, vec[i].len);
+        assert(ret < 0);
```

Comment:
assert is there to replicated the original assert in the code, it wasn't a
1:1 transalation so I may had messed this one, that is the point to have
more ppl testing it. If it is to remove it, ok, but ret should not be
negative, so the assert was helpful. Can it be handled differently in this
case? Dunno, you tell me, but it looks like that return failure here will
just trigger a similar action up in the code, dunno fore sure yet.

@Tom, sadly the auto was not to blame in this case since 'ret' is typed :-P

On Thu, Aug 4, 2016 at 11:59 AM, Georgii Staroselskii <
notifications@github.com> wrote:

> In libraries/AP_HAL_Linux/UARTDriver.cpp
> <https://github.com/ArduPilot/ardupilot/pull/4593#discussion_r73538694>:
>
> > -            _read_fd(&_readbuf[_readbuf_tail], n);
> > -        } else {
> > -            assert(_readbuf_tail+n1 <= _readbuf_size);
> > -            int ret = _read_fd(&_readbuf[_readbuf_tail], n1);
> > -            if (ret == n1 && n > n1) {
> > -                assert(_readbuf_tail+(n-n1) <= _readbuf_size);
> > -                _read_fd(&_readbuf[_readbuf_tail], n - n1);
> > -            }
> > -        }
> > +    int ret;
> > +    ByteBuffer::IoVec vec[2];
> > +
> > +    const auto n_vec = _readbuf.reserve(vec, _readbuf.space());
> > +    for (int i = 0; i < n_vec; i++) {
> > +        ret = _read_fd(vec[i].data, vec[i].len);
> > +        assert(ret < 0);
>
> Although the assert seems wrong, the removal of it doesn't really help.
> The console is getting polluted with messages like
> Info: GPS 0: detected as NMEA at 9600 baud
> Info: GPS 0: detected as u-blox at 115200 baud
> (the list goes on and on)
>
> So that was kind of a good thing we had this assert in there.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ArduPilot/ardupilot/pull/4593/files/d87284bccc3713ffa27310e16b1294cdc76b50ee#r73538694>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABgsZYgT4cq_Cf6RU2heMlIY_0znOqkWks5qcf5MgaJpZM4JaxPM>
> .
>